datasets/machine-learning-dataset-for-sdo/sdo_ml_indexer.py

```diff
@@ -54,32 +69,166 @@ def index(data_dir):
+                                freq="MS").to_pydatetime().tolist()
+
+    end_dates = pd.date_range(start=adjusted_start, end=adjusted_end,
+                              freq="M").to_pydatetime().tolist()
+
+    if len(start_dates) == 0:
+        return [(adjusted_start, adjusted_end)]
+
+    ranges = []
+    for s, e in zip(start_dates, end_dates):
+        ranges.append((s, e))
+
+    if len(ranges) == 0:
+        return [(adjusted_start, adjusted_end)]
+
+    return ranges
+
+
+goes_cache = {}
+
+
+def fetch_goes_metadata(start, end, path):
+    # NOTE it can happen that sunpy will download invalid files during rate limiting, find them using grep -r "Too many" ~/sunpy/data/*
+    # https://docs.sunpy.org/en/v3.0.0/generated/gallery/acquiring_data/goes_xrs_example.html
+    # https://www.ngdc.noaa.gov/stp/satellite/goes-r.html
+    for t_start, t_end in get_date_ranges(start, end):
+        goes_dir = Path(path) / Path("goes_cache") / Path(str(t_start.year)) /\
+            Path(str(t_start.month))
+        goes_path = goes_dir / Path('goes_ts.csv')
+        if not pathlib.Path.exists(goes_path):
+            try:
+                print(f"retrieving goes data for {goes_path}")
+                os.makedirs(goes_dir, exist_ok=True)
+                satellite = get_goes_satellite(t_start)
+                search_result = Fido.search(
+                    a.Time(t_start, t_end), a.Instrument.xrs, satellite)
+                download_result = Fido.fetch(search_result)
+                goes_ts = ts.TimeSeries(download_result)
+                if isinstance(goes_ts, list) and len(goes_ts) > 0:
+                    frames = []
+                    for goes_ts_frm in goes_ts:
+                        frames.append(goes_ts_frm.to_dataframe())
+                    goes_ts_df = pd.concat(frames)
+                else:
+                    goes_ts_df = goes_ts.to_dataframe()
+
+                goes_ts_df.index.name = 'timestamp'
+                goes_ts_df.to_csv(goes_path, index=True,
+                                  date_format=ISO_FORMAT)
+            except Exception as e:
+                print(e)
+                goes_ts_df = pd.DataFrame()
+        else:
+            try:
+                print(f"goes data for {goes_path} already present")
+                if goes_path in goes_cache:
+                    goes_ts_df = goes_cache[goes_path]
+                else:
+                    goes_ts_df = pd.read_csv(goes_path)
+                    goes_ts_df["timestamp"] = pd.to_datetime(
+                        goes_ts_df['timestamp'], format=ISO_FORMAT)
+                    goes_ts_df = goes_ts_df.set_index('timestamp', drop=False)
+                    goes_ts_df = goes_ts_df.sort_index()
+                    goes_cache[goes_path] = goes_ts_df
+
+            except Exception as e:
+                print(e)
+                goes_ts_df = pd.DataFrame()
+
+    return goes_ts_df
+
+
+def get_goes_satellite(date):
```

Comment:
this function will determine which satellite to use, possibly a more advanced algorithm could be used
<img width="910" alt="image" src="https://user-images.githubusercontent.com/14012777/153555971-6c210ca2-7306-4fd4-b3e2-fe03a27442ee.png">